_posts/cleaning-up-your-test-data-builders.md

```diff
@@ -0,0 +1,104 @@
+---
+
+I've been using the [Test Data Builder](http://www.natpryce.com/articles/000714.html) pattern for quite a while now. It's really improved the readability of my test automation code. I've always had the nagging feeling though that it could be better.
+
+Inspired by a talk I attended a couple of months ago I decided to refactor some of my test data builders. In this post I'll go through some of the steps I've taken.
+
+The code example below is my initial take on a builder to create a simple Person object. It has a name, gender and address.
+
+```csharp
+Person sherlock = new PersonBuilder()
+    .WithName("Sherlock Holmes")
+    .IsMan()
+    .WithAddress(
+        new AddressBuilder()
+            .WithAddressLine1("221B Baker Street")
+            .WithAddressLine2("London")
+            .Build()
+    )
+    .Build();
+```
+
+### Introducing the Object Mother
+
+Step one is to move the instantiation of the `PersonBuilder`. In the test I'm not really interested in the fact that I'm using a builder. I want 'A Person'. We can combine the [Object Mother](https://martinfowler.com/bliki/ObjectMother.html) pattern with the Test Data Builder pattern to make this happen.
+
+I've created an object mother class called `A`. It has a static `Person` property that returns a new instance of the `PersonBuilder`.  
+
+```csharp
+Person sherlock = A.Person
+    .WithName("Sherlock Holmes")
+    .IsMan()
+    .WithAddress(
+        new AddressBuilder()
+            .WithAddressLine1("221B Baker Street")
+            .WithAddressLine2("London")
+            .Build()
+    )
+    .Build();
+```
+
+### Moving a build method to the Object Mother
+
+We can take this one step further and create a `Man` property in the object mother. Combining `A.Person` and the `IsMan` method into one.
+
+```csharp
+Person sherlock = A.Man
+    .WithName("Sherlock Holmes")
+    .WithAddress(
+        new AddressBuilder()
+            .WithAddressLine1("221B Baker Street")
+            .WithAddressLine2("London")
+            .Build()
+    )
+    .Build();
+```
+
+Be cautious with this. Don't use more than one (business) concept like `A.ManWithName("Sherlock Holmes")`. Combining multiple build steps is what the builder is for.
+
+### Rename methods to improve flow
+
+The next step is to improve the flow of the code. I've renamed `WithName` to `Called` and `WithAddress` to `LivingAt`. Making the code read more like a normal sentence.
+
+```csharp
+Person sherlock = A.Man.Called("Sherlock Holmes")
+    .LivingAt(
+        new AddressBuilder()
+            .WithAddressLine1("221B Baker Street")
+            .WithAddressLine2("London")
+            .Build()
+    )
+    .Build();
+```
+
+### Simplifying address creation
+
+The `AddressBuilder` is still a bit ugly. So I've refactored the `LivingAt` method to take the first and second address line in one string, separated by a comma. The `AddressBuilder` is used inside `LivingAt` to parse the string.  
```

Comment:
The AddressBuilder is not really ugly since we never see it's implementation. I would say that the usage of the Address builder inside your LivingAt method looks ugly.